includes/payment-tokens/class-wc-stripe-klarna-payment-token.php

```diff
@@ -0,0 +1,120 @@
+ *
+ * @package WooCommerce_Stripe
+ * @since 10.0.0
+ */
+
+// phpcs:disable WordPress.Files.FileName
+
+// Exit if accessed directly
+defined( 'ABSPATH' ) || exit;
+
+class WC_Payment_Token_Klarna extends WC_Payment_Token implements WC_Stripe_Payment_Method_Comparison_Interface {
+	/**
+	 * Token Type.
+	 *
+	 * @var string
+	 */
+	protected $type = WC_Stripe_Payment_Methods::KLARNA;
+
+	/**
+	 * Extra data.
+	 *
+	 * @var string[]
+	 */
+	protected $extra_data = [
+		'dob' => '',
+	];
+
+	/**
+	 * Returns the name of the token to display
+	 *
+	 * @param  string $deprecated Deprecated since WooCommerce 3.0
+	 * @return string The name of the token to display
+	 */
+	public function get_display_name( $deprecated = '' ) {
+		return __( 'Klarna', 'woocommerce-gateway-stripe' );
+	}
+
+	/**
+	 * Sets the Klarna token's date of birth.
+	 *
+	 * @param string $dob The formatted date of birth object (YYYY-mm-dd).
+	 */
+	public function set_dob( string $dob ) {
+		$this->set_prop( 'dob', $dob );
+	}
+
+	/**
+	 * Fetches Klarna token's date of birth (formatted to Y-m-d).
+	 *
+	 * @return string The Klarna token's date of birth.
+	 */
+	public function get_dob(): string {
+		$dob = $this->get_prop( 'dob' );
+		if ( is_string( $dob ) ) {
+			return $dob;
+		}
+		return '';
+	}
+
+	/**
+	 * Sets the Klarna token's date of birth based on the object returned from the Stripe payment method API.
+	 *
+	 * @param object $dob The raw `dob` object from Stripe.
+	 * @see https://docs.stripe.com/api/payment_methods/object#payment_method_object-klarna-dob
+	 */
+	public function set_dob_from_object( object $dob ) {
+		$this->set_dob( $this->format_dob( $dob ) );
+	}
+
+	/**
+	 * Checks if the payment method token is equal a provided payment method.
+	 *
+	 * @inheritDoc
+	 */
+	public function is_equal_payment_method( $payment_method ): bool {
+		if ( WC_Stripe_Payment_Methods::KLARNA !== $payment_method->type ) {
+			return false;
+		}
+
+		$payment_method_dob = $payment_method->klarna->dob ?? null;
+		if ( empty( $this->get_dob() ) ) {
+			return null === $payment_method_dob;
+		}
+
+		if ( null === $payment_method_dob || empty( get_object_vars( $payment_method_dob ) ) ) {
+			return false;
+		}
+
+		return $this->format_dob( $payment_method_dob ) === $this->get_dob();
+	}
+
+	/**
+	 * Returns this token's hook prefix.
+	 *
+	 * @return string The hook prefix.
+	 */
+	protected function get_hook_prefix() {
+		return 'woocommerce_payment_token_klarna_get_';
+	}
+
+	/**
+	 * Formats the date of birth for display.
+	 *
+	 * @param $dob object The date of birth object.
+	 * @return string The formatted date of birth.
+	 */
+	protected function format_dob( object $dob ) {
+		$dob_parts = [
+			$dob->year ?? null,
+			isset( $dob->month ) ? str_pad( $dob->month, 2, '0', STR_PAD_LEFT ) : null,
+			isset( $dob->day ) ? str_pad( $dob->day, 2, '0', STR_PAD_LEFT ) : null,
+		];
+		return implode( '-', array_filter( $dob_parts ) );
```

Comment:
This implementation feels like it will break down for edge cases. As an example, if we get one `dob` with only a day value of `10` and another dob object with only a month of `10`, the resulting strings for both objects would be `10`.

I think it would be more robust to guard against all empty values first, and then build a string with placeholders for unknown values, as that should make eventual string comparisons super-clear.
 
```suggestion
		if ( empty( $dob->year ) && empty( $dob->month ) && empty( $dob->day ) ) {
			// Either return null or ''. Empty string probably make more sense.
			return '';
		}
		$dob_parts = [
			$dob->year ?? 'YYYY',
			$dob->month ? str_pad( $dob->month, 2, '0', STR_PAD_LEFT ) : 'MM',
			$dob->day ? str_pad( $dob->day, 2, '0', STR_PAD_LEFT ) : 'DDDD',
		];
		return implode( '-', array_filter( $dob_parts ) );
```